aeneas/src/mach/MachProgram.v3

```diff
@@ -193,15 +194,22 @@ class MachProgram extends TargetProgram {
 	}
 	def machVal(val: Val) -> Val {
 		if (val == null) return null;
-		if (Record.?(val)) {
-			var r = Record.!(val);
-			if (V3.isVariant(r.rtype)) {
-				// check for variants represented as enums
-				if (prog.ir.isEnum(r.rtype)) return Int.box(V3.getVariantTag(r.rtype));
+		match (val) {
+			x: Record => {
+				if (V3.isVariant(x.rtype)) {
+					// check for variants represented as enums
+					if (prog.ir.isEnum(x.rtype)) return Int.box(V3.getVariantTag(x.rtype));
+				}
+				return addrOfRecord(x);
+			}
+			x: FuncVal => return addrOfMethod(x.memberRef.asMethod());
+			x: ArrayRangeStart => {
+				if (x.arrayType == null) return Int.box(x.start);
```

Comment:
How does it happen that the `arrayType` is null here?

Reply:
If a range has a default value, and we later use that default RangeStart (`null`) and do a `RangeStartPlusIndex` on it, the `RangeStartPlusIndex` won't know what `arrayType` to put in the constructed `ArrayRangeStart`.